File: problems/problems_2181/solution.ts
import {ListNode,IntArrayToLinkedList,LinkedListToIntArray} from "../../typescript/models/listnode";

/**
 * Definition for singly-linked list.
 * class ListNode {
 *     val: number
 *     next: ListNode | null
 *     constructor(val?: number, next?: ListNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.next = (next===undefined ? null : next)
 *     }
 * }
 */

function mergeNodes(head: ListNode | null): ListNode | null {
	const dummy: ListNode | null = new ListNode(0);
	let node: ListNode | null = dummy, current: ListNode | null = head.next;
	while (current !== null) {
		let s: number = 0;
		while (current !== null && current.val !== 0) {
			s += current.val;
			current = current.next;
		}
		node.next = new ListNode(s);
		node = node.next;
		if (current !== null) {
			current = current.next;
		}
	}
	return dummy.next;
};

export function Solve(inputJsonElement: string): any {
	const inputValues: string[] = inputJsonElement.split("\n");
	const head: ListNode | null = IntArrayToLinkedList(JSON.parse(inputValues[0]));
	return LinkedListToIntArray(mergeNodes(head));
}

File: problems/problems_2181/solution.go
package problem2181

import (
	"encoding/json"
	. "leetCode/golang/models"
	"log"
	"strings"
)

/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func mergeNodes(head *ListNode) *ListNode {
	dummy := &ListNode{Val: 0}
	node, cur := dummy, head.Next
	for cur != nil {
		s := 0
		for cur != nil && cur.Val != 0 {
			s += cur.Val
			cur = cur.Next
		}
		node.Next = &ListNode{Val: s}
		node = node.Next
		if cur != nil {
			cur = cur.Next
		}
	}
	return dummy.Next
}

func Solve(inputJsonValues string) any {
	inputValues := strings.Split(inputJsonValues, "\n")
	var head *ListNode

	var headIntArray []int
	if err := json.Unmarshal([]byte(inputValues[0]), &headIntArray); err != nil {
		log.Fatal(err)
	}
	head = IntArrayToLinkedList(headIntArray)

	return LinkedListToIntArray(mergeNodes(head))
}
